### ISSUE TITLE
NotNullIfNotNull also requires casting to non-nullable type for valuetypes

### ISSUE THREAD

**Issue by codetuner** [dotnet/roslyn#71728]
Take the following code in a Console App:

```
namespace NotNulIfNotNullEnumRepro
{
    public enum Colors
    {
        Black = 0,
        White = 1
    }

    public enum Couleurs
    {
        Noir = 0,
        Blanc = 1
    }

    public class English
    {
        public string Name { get; set; } = String.Empty;
        public int Age { get; set; }
        public Colors Color { get; set; }
    }

    public class Francais
    {
        public string Nom { get; set; } = String.Empty;
        public int Age { get; set; }
        public Couleurs Couleur { get; set; }
    }

    internal class Program
    {
        static void Main(string[] args)
        {
            var e = new English();
            var f = new Francais();

            f.Nom = (string)TranslateString(e.Name);        // Line 36
            f.Age = (int)TranslateAge(e.Age);               // Line 37
            f.Couleur = (Couleurs)TranslateColor(e.Color);  // Line 38
        }

        //[return: System.Diagnostics.CodeAnalysis.NotNullIfNotNull("str")]
        public static string? TranslateString(string? str)
        {
            throw new NotImplementedException();
        }

        //[return: System.Diagnostics.CodeAnalysis.NotNullIfNotNull("val")]
        public static int? TranslateAge(int? val)
        {
            throw new NotImplementedException();
        }

        //[return: System.Diagnostics.CodeAnalysis.NotNullIfNotNull("col")]
        public static Couleurs? TranslateColor(Colors? col)
        {
            throw new NotImplementedException();
        }
    }
}
```

It does compile but gives the following warnings:
```
Program.cs(36,21,36,52): warning CS8600: Converting null literal or possible null value to non-nullable type.
Program.cs(36,21,36,52): warning CS8601: Possible null reference assignment.
Program.cs(37,21,37,45): warning CS8629: Nullable value type may be null.
Program.cs(38,25,38,58): warning CS8629: Nullable value type may be null.
```

Uncomment the 3 "//[return: System.Diagnostics.CodeAnalysis.NotNullIfNotNull..." lines and it compiles without warnings.
So the NotNullIfNotNull attribute seems to do its work.

But remove the cast operators on lines 36, 37 and 38 and it fails to compile with following errors:

```
Program.cs(37,21,37,40): error CS0266: Cannot implicitly convert type 'int?' to 'int'. An explicit conversion exists (are you missing a cast?)
Program.cs(38,25,38,48): error CS0266: Cannot implicitly convert type 'NotNulIfNotNullEnumRepro.Couleurs?' to 'NotNulIfNotNullEnumRepro.Couleurs'. An explicit conversion exists (are you missing a cast?)
```

It only complains about the Int32 and Enum type parameters.

In conclusion:
- casting without NotNullIfNotNull attribute compiles but results in warnings
- NotNullIfNotNull attribute without casting does not compile
Hence, both casting and the NotNullIfNotNull attribute are required, which sounds like the compiler is not trusting the NotNullIfNotNull attribute...

**Comment by chrisoverzero** [comment]
> It only complains about the `Int32` and `Enum` type parameters.

Yes, because those are value types.

When a reference type’s name is followed by a question mark (as in `string?`) it’s still a `string`, but the value’s nullability is tracked by the compiler. A value of type `string?` believed not to be `null` can be assigned to `string`. When a value type’s name is followed by a question mark (as in `int?`) it is a `Nullable<int>`. This is a completely different type from `int`, so a value of its type cannot be directly assigned to an `int` irrespective of nullability tracking.

**Comment by jaredpar** [comment]
@chrisoverzero analysis is correct here. This is the expected behavior of the compiler.
